runtime-sdk-macros/src/module_derive/migration_handler.rs

```diff
@@ -0,0 +1,167 @@
+                self.genesis_ty = Some(item);
+
+                None // Take the item.
+            }
+            syn::ImplItem::Fn(ref f) => {
+                // Check whether a `migration` attribute is set for the method.
+                if let Some(attrs) = parse_attrs(&f.attrs) {
+                    self.migrate_fns.push(MigrateFn {
+                        ident: f.sig.ident.clone(),
+                        from_version: attrs.from_version,
+                        item,
+                    });
+
+                    None
+                } else {
+                    Some(item) // Return the item.
+                }
+            }
+            _ => Some(item), // Return the item.
+        }
+    }
+
+    fn derive(&mut self, generics: &syn::Generics, ty: &Box<syn::Type>) -> TokenStream {
+        let genesis_ty = if let Some(genesis_ty) = &self.genesis_ty {
+            genesis_ty
+        } else {
+            return quote! {};
+        };
+
+        let mut seen_versions = HashSet::new();
+        let (migrate_fns, mut migrate_arms): (Vec<_>, Vec<_>) = self.migrate_fns.iter().map(|f| {
+            let MigrateFn { item, ident, from_version } = f;
+            if seen_versions.contains(from_version) {
+                emit_compile_error(format!(
+                    "Duplicate migration for version: {from_version}"
+                ));
+            }
+            seen_versions.insert(from_version);
+
+            (
+                item,
+                if from_version == &0 {
+                    // Version zero is special as initializing from genesis always gets us latest.
+                    quote! { if version == #from_version { Self::#ident(genesis); version = Self::VERSION; } }
+                } else {
+                    // For other versions, each migration brings us from V to V+1.
+                    quote! { if version == #from_version { Self::#ident(); version += 1; } }
+                }
+            )
+        }).unzip();
+
+        // Ensure there is a genesis migration, at least an empty one that bumps the version.
+        if !seen_versions.contains(&0) {
+            migrate_arms.push(quote! {
+                if version == 0u32 { version = Self::VERSION; }
+            });
+        }
+
+        quote! {
+            #[automatically_derived]
+            impl #generics sdk::module::MigrationHandler for #ty {
+                #genesis_ty
+
+                fn init_or_migrate<C: Context>(
+                    _ctx: &mut C,
+                    meta: &mut sdk::modules::core::types::Metadata,
+                    genesis: Self::Genesis,
+                ) -> bool {
+                    let mut version = meta.versions.get(Self::NAME).copied().unwrap_or_default();
+                    if version == Self::VERSION {
+                        return false; // Already the latest version.
+                    }
+
+                    #(#migrate_arms)*
+
+                    if version != Self::VERSION {
```

Comment:
Is there a problem if one defines migration from `Self::VERSION` to `Self::VERSION + 1`? Should we exclude this migration?

Reply:
Ideally we could exclude this at compile time, but unfortunately it seems this doesn't seem straightforward due to `Self` use in const functions.